### PR TITLE
crimson/osd: introduce the InternalClientRequest infrastructure.

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test api

File: src/crimson/osd/object_context.h [hunk 1]
```diff
@@ -49,7 +49,7 @@ class ObjectContext : public ceph::common::intrusive_lru_base<
   using watch_key_t = std::pair<uint64_t, entity_name_t>;
   std::map<watch_key_t, seastar::shared_ptr<crimson::osd::Watch>> watchers;
 
-  ObjectContext(const hobject_t &hoid) : obs(hoid) {}
+  ObjectContext(hobject_t hoid) : obs(std::move(hoid)) {}
```
Comment (tchaikov) [inline src/crimson/osd/object_context.h:52]: `ObjectContextRegistry::get_cached_obc(const hobject_t &hoid)` and ``intrusive_lru::get_or_create()` still take a const reference.  so we are not able to *move* the `hoid` down to `ObjectContext()` without changing all the functions on the call stack.

File: src/crimson/osd/pg.cc [hunk 2]
```diff
@@ -1007,6 +1040,26 @@ PG::with_locked_obc(Ref<MOSDOp> &m, const OpInfo &op_info,
   };
 }
 
+template <RWState::State State>
+PG::interruptible_future<>
+PG::with_locked_obc(ObjectContextRef obc, PG::with_obc_func_t &&f)
+{
+  // TODO: a question from rebase: do we really need such checks when
+  // the interruptible stuff is being used?
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc:1075]: @xxhdx1985126 can we drop this check?

File: src/crimson/osd/pg.cc [hunk 3]
```diff
@@ -857,13 +857,10 @@ std::optional<hobject_t> PG::resolve_oid(
...
-PG::with_head_obc(hobject_t oid, with_obc_func_t&& func)
+PG::with_head_obc(ObjectContextRef obc, bool existed, with_obc_func_t&& func)
 {
-  logger().debug("{} {}", __func__, oid);
-  boost::intrusive_ptr<PG> pgref{this};
-  assert(oid.is_head());
-  auto [obc, existed] =
-    shard_services.obc_registry.get_cached_obc(std::move(oid));
+  logger().debug("{} {}", __func__, obc->get_oid());
+  assert(obc.is_head());
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc:906]: `obc` is a smart pointer, so i think we should put
```c+++
assert(obc->is_head());
```
here instead.

File: src/crimson/osd/osd_operations/internal_client_request.cc [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                      std::move(osd_ops),
+                      std::as_const(op_info),
+                      get_do_osd_ops_params(),
+                      [this] {
+                        return PG::do_osd_ops_iertr::now();
+                      },
+                      [this] (const std::error_code& e) {
+                        return PG::do_osd_ops_iertr::now();
+                      }
+                    ).safe_then_interruptible(
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/internal_client_request.cc:83]: Move the `ct_error::eagain` handling below and restart the loop.

File: src/crimson/osd/osd_operations/internal_client_request.cc [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                        return interruptor::now();
+                      }, crimson::ct_error::eagain::handle([this]() mutable {
+                        return interruptor::now();
+                      })
+                    );
+                  });
+                });
+              });
+            }).handle_error_interruptible(PG::load_obc_ertr::all_same_way([] {
+              return seastar::now();
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/internal_client_request.cc:94]: Implement the error handling.

File: src/crimson/osd/pg.cc (hunk 3 above)
Comment (rzarzynski) [reply src/crimson/osd/pg.cc]: Thanks for catching this! I was using  a production build.

Now should be fixed.

File: src/crimson/osd/pg.cc [hunk 6]
```diff
@@ -693,90 +691,138 @@ PG::interruptible_future<> PG::repair_object(
...
-    return repair_object(m, obc->obs.oi.soid, obc->obs.oi.version)
-    .then_interruptible([] {
-      return PG::do_osd_ops_ertr::future<Ref<MOSDOpReply>>(
-      crimson::ct_error::eagain::make());
     });
-  }), osd_op_errorator::all_same_way([ox = std::move(ox),
-                                      m,
-                                      obc,
-                                      this] (const std::error_code& e) {
-    return handle_failed_op(e, std::move(obc), *ox, *m);
```
Comment (rzarzynski) [inline src/crimson/osd/pg.cc:776]: It's quite probable `handle_failed_op()` can be dropped now.

File: src/crimson/osd/pg.cc [hunk 7]
```diff
@@ -878,12 +919,31 @@ PG::with_head_obc(hobject_t oid, with_obc_func_t&& func)
...
+PG::with_head_obc(hobject_t oid, with_obc_func_t&& func)
+{
+  auto [obc, existed] =
+    shard_services.obc_registry.get_cached_obc(std::move(oid));
+  return with_head_obc<State>(std::move(obc), existed, std::move(func));
+}
+
+template<RWState::State State>
+PG::interruptible_future<>
+PG::with_existing_head_obc(ObjectContextRef obc, with_obc_func_t&& func)
```
Comment (rzarzynski) [inline src/crimson/osd/pg.cc:939]: The `obc`-taking variants of `with_..._obc()` look like a dead end of the evolution. Ultimately `InternalClientRequest` reverted to the plain variant which always succeeds retrieving the context from cache as it's pinned with an extra `ObjectContextRef` instance. Less efficient but simpler that way.

File: src/crimson/osd/ops_executer.h [hunk 8]
```diff
@@ -269,15 +320,17 @@ auto OpsExecuter::with_effect_on_obc(
...
+OpsExecuter::flush_changes_n_do_ops_effects(Ref<PG> pg, MutFunc&& mut_func) &&
 {
   const bool want_mutate = !txn.empty();
   // osd_op_params are instantiated by every wr-like operation.
   assert(osd_op_params || !want_mutate);
   assert(obc);
   auto maybe_mutated = interruptor::make_interruptible(osd_op_errorator::now());
   if (want_mutate) {
+    osd_op_params->req_id = msg->get_reqid();
+    //osd_op_params->mtime = msg->get_mtime();
```
Comment (rzarzynski) [inline src/crimson/osd/ops_executer.h:333]: Oops, TODO.

File: src/crimson/osd/object_context.h (hunk 1 above)
Comment (rzarzynski) [reply src/crimson/osd/object_context.h:52]: Indeed, this change eliminates only the `ObjectContext` blocker towards the copy-free initialization. We can go down the stack in a follow-up.

**Comment by tchaikov** [comment]
@rzarzynski in the light of merged #40646 , can we close this PR ?

**Comment by rzarzynski** [comment]
Yup, all parts of this have been merged with the watch timeout PR. Closing.

**Pull request closed without merging**